java/src/main/java/org/apache/arrow/gandiva/evaluator/Types.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (C) 2017-2018 Dremio Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.gandiva.evaluator;
+
+import com.google.common.collect.Lists;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+import org.apache.arrow.gandiva.exceptions.GandivaException;
+import org.apache.arrow.gandiva.ipc.GandivaTypes;
+import org.apache.arrow.gandiva.ipc.GandivaTypes.ExtGandivaType;
+import org.apache.arrow.gandiva.ipc.GandivaTypes.GandivaDataTypes;
+import org.apache.arrow.gandiva.ipc.GandivaTypes.GandivaFunctions;
+import org.apache.arrow.gandiva.ipc.GandivaTypes.GandivaType;
+import org.apache.arrow.vector.types.DateUnit;
+import org.apache.arrow.vector.types.FloatingPointPrecision;
+import org.apache.arrow.vector.types.TimeUnit;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+
+import java.util.List;
+
+/**
+ * Used to get the functions and data types supported by
+ * Gandiva.
+ * All types are in Arrow namespace.
+ */
+public class Types {
+
+  private static final int BIT_WIDTH8 = 8;
+  private static final int BIT_WIDTH_16 = 16;
+  private static final int BIT_WIDTH_32 = 32;
+  private static final int BIT_WIDTH_64 = 64;
+  private static final boolean IS_SIGNED_FALSE = false;
+  private static final boolean IS_SIGNED_TRUE = true;
+
+  private final List<ArrowType> supportedTypes;
+  private final List<FunctionSignature> functionSignatures;
+
+  private static volatile Types INSTANCE;
+
+  private Types(List<ArrowType> supportedTypes, List<FunctionSignature> functionSignatures) {
+    this.supportedTypes = supportedTypes;
+    this.functionSignatures = functionSignatures;
+  }
+
+  /**
+   * Returns a singleton instance of the class.
+   * @return singleton instance
+   * @throws GandivaException if error in Gandiva Library integration.
+   */
+  public static Types getInstance() throws GandivaException {
+    if (INSTANCE == null) {
+      synchronized (Types.class) {
+        if (INSTANCE == null) {
+          // ensure library is setup.
+          NativeBuilder.getInstance();
+          List<ArrowType> typesFromGandiva = getSupportedTypesFromGandiva();
+          List<FunctionSignature> functionsFromGandiva = getSupportedFunctionsFromGandiva();
+          INSTANCE = new Types(typesFromGandiva, functionsFromGandiva);
+        }
+      }
+    }
+    return INSTANCE;
+  }
+
+  public List<FunctionSignature> getSupportedFunctions() {
```

Comment:
Why not make these functions private and support a method isFunctionSupported(name, array of args, return type)?

This allows us to change our internal implementation without requiring Gandiva users to understand FunctionSignature, etc.

Reply:
That makes the communication too chatty and we want to integrate the Gandiva Registry with the Dremio registry which requires this (going forward we would have gandiva only functions which will not be in the dremio registry thus requiring the set to be exported..)